src/components/Dialogs/QueueModal.tsx

```diff
@@ -0,0 +1,234 @@
+
+  const items = useMemo(() => {
+    // Don't render items if the modal is not shown.
+    if (!showModal) {
+      return [];
+    }
+
+    if (expectedTab !== activeTab) {
+      if (query.isUninitialized || query.isLoading || query.isFetching) {
+        return [];
+      }
+      setExpectedTab(activeTab);
+    }
+
+    let array = query.data?.List ?? [];
+    if (currentCommand != null) {
+      array = filter(array, item => item.ID !== currentCommand.ID);
+      // Remove the last item if the above filtering did nothing.
+      if (pageSize !== 0 && array.length >= pageSize) {
+        array = array.slice(0, -1);
+      }
+    }
+
+    const itemArray = map(array, item => (
+      <div className="mt-2 flex gap-x-3" key={`item-${item.ID}`}>
+        <div className="flex-grow">
+          {item.Name}
+        </div>
+        <div className={cx(['px-4', item.IsRunning ? 'text-panel-important' : undefined, item.IsDisabled ? 'text-panel-warning' : undefined])}>
+          {/* eslint-disable-next-line no-nested-ternary */}
+          <Icon path={item.IsRunning ? mdiRun : item.IsDisabled ? mdiAlertCircleOutline : mdiHelpCircleOutline} size={1} />
+        </div>
+      </div>
+    ));
+    if (currentCommand != null) {
+      itemArray.unshift(
+        <div className="mt-2 flex gap-x-3" key={`item-${currentCommand.ID}`}>
+          <div className="flex-grow">
+            {currentCommand.Name}
+          </div>
+          <div className={cx(['px-4', currentCommand.IsRunning ? 'text-panel-important' : undefined])}>
+            <Icon path={currentCommand.IsRunning ? mdiRun : mdiHelpCircleOutline} size={1} />
+          </div>
+        </div>,
+      );
+    }
+    return itemArray;
+  }, [showModal, activeTab, query.data, pageSize, currentCommand, expectedTab, query.isUninitialized, query.isLoading, query.isFetching]);
+
+  const handlePageSizeChange = useEventCallback((e: React.ChangeEvent<HTMLInputElement>) => {
+    e.preventDefault();
+    let value = parseInt(e.target.value, 10);
+    if (Number.isNaN(value)) {
+      return;
+    }
+
+    if (!Number.isInteger(value)) {
+      value = Math.floor(value);
+    }
+
+    // eslint-disable-next-line no-nested-ternary
+    value = value > 100 ? 100 : value < 0 ? 0 : value;
+
+    setPageSize(value);
+  });
+
+  const handleShowDisabledToggle = useEventCallback(() => {
+    setShowAll(!showAll);
+  });
+
+  const handleToggleAllQueues = useEventCallback(async () => {
+    await queueOperation({ operation: isAllPaused ? 'StartAll' : 'StopAll' });
+  });
+
+  const handleToggleQueue = useEventCallback(async () => {
+    await queueOperation({ operation: isPaused ? 'Start' : 'Stop', queue: activeTab });
+  });
+
+  // We're intentionally not letting RTK invalidate any tags for automagic query
+  // updates, since if we did then it would fetch in the background when the
+  // modal is not in view, since it's technically always mounted, just not
+  // shown.
+  useLayoutEffect(() => {
```

Comment:
why `useLayoutEffect` instead of `useEffect`?

Reply:
Because otherwise it will render before it should. E.g. it shows the empty queue page, then the loader, then the items, instead of just showing the loader then the items, also the same issues while switching tabs. If you have a better way that also works with `useEffect`  instead of `useLayoutEffect` then I'm all ears.

Reply:
I'm confused... useLayoutEffect runs *before* useEffect

Reply:
Also, we're talking in terms of milliseconds, or a flicker on the screen, when i say it's showing the empty queue page (when it shouldn't).

Reply:
useLayoutEffect is for libraries as far as I remember, also docs say 

> The code inside useLayoutEffect and all state updates scheduled from it block the browser from repainting the screen. When used excessively, this makes your app slow. When possible, prefer [useEffect.](https://react.dev/reference/react/useEffect)